arch/xtensa/core/mpu.c

```diff
@@ -701,3 +707,419 @@ void xtensa_mpu_init(void)
+	 * running thread.
+	 */
+	if (thread == _current_cpu->current) {
+		xtensa_mpu_map_write(thread->arch.mpu_map);
+	}
+
+	return ret;
+}
+
+int arch_mem_domain_thread_remove(struct k_thread *thread)
+{
+	uintptr_t stack_end_addr;
+	int ret;
+
+	struct k_mem_domain *domain = thread->mem_domain_info.mem_domain;
+
+	if ((thread->base.user_options & K_USER) == 0) {
+		ret = 0;
+		goto out;
+	}
+
+	if ((thread->base.thread_state & _THREAD_DEAD) == 0) {
+		/* Thread is migrating to another memory domain and not
+		 * exiting for good; we weren't called from
+		 * z_thread_abort().  Resetting the stack region will
+		 * take place in the forthcoming thread_add() call.
+		 */
+		ret = 0;
+		goto out;
+	}
+
+	stack_end_addr = thread->stack_info.start + thread->stack_info.size;
+	if (stack_end_addr < thread->stack_info.start) {
+		/* Account for wrapping around back to 0. */
+		stack_end_addr = 0xFFFFFFFFU;
+	}
+
+	/*
+	 * Restore permissions on the thread's stack area since it is no
+	 * longer a member of the domain.
+	 */
+	ret = mpu_map_region_add(&domain->arch.mpu_map,
+				 thread->stack_info.start, stack_end_addr,
+				 XTENSA_MPU_ACCESS_P_RW_U_NA,
+				 CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE,
+				 NULL);
+
+	xtensa_mpu_map_write(thread->arch.mpu_map);
+
+out:
+	return ret;
+}
+
+int arch_buffer_validate(void *addr, size_t size, int write)
+{
+	uintptr_t start_addr, end_addr;
+	struct xtensa_mpu_entry *entry_slot_s, *entry_slot_e;
+	bool exact_s, exact_e;
+	uint8_t idx_s, idx_e, first_enabled_idx;
+
+	int ret = 0;
+	const struct k_thread *thread = _current;
+	const struct xtensa_mpu_map *map = thread->arch.mpu_map;
+	const struct xtensa_mpu_entry *entries = map->entries;
+
+	/* Make sure the start address is aligned. */
+	start_addr = ROUND_DOWN((uintptr_t)addr, XCHAL_MPU_ALIGN);
+
+	/* Figure out the aligned ending address. */
+	end_addr = (uintptr_t)addr + size;
+	if (end_addr < (uintptr_t)addr) {
+		/* Assume end of memory if wrap around. */
+		end_addr = 0xFFFFFFFFU;
+	} else {
+		end_addr = ROUND_UP(end_addr, XCHAL_MPU_ALIGN);
+	}
+
+	first_enabled_idx = find_first_enabled_entry(entries);
+
+	/* Find the slot where each address belongs to. */
+	entry_slot_s = (struct xtensa_mpu_entry *)
+		       check_addr_in_mpu_entries(entries, start_addr, first_enabled_idx,
+						 &exact_s, &idx_s);
+	entry_slot_e = (struct xtensa_mpu_entry *)
+		       check_addr_in_mpu_entries(entries, end_addr, first_enabled_idx,
+						 &exact_e, &idx_e);
+
+	/* NULL means the address is not in map at all. Bail. */
+	if ((entry_slot_s == NULL) || (entry_slot_e == NULL)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/*
+	 * If ending address does not match the slot address exactly,
+	 * we will need to consider this entry for permission checking.
+	 */
+	if (!exact_e) {
+		idx_e++;
+	}
+
+	for (int i = idx_s; i < idx_e; i++) {
+		uint8_t access_rights = xtensa_mpu_entry_access_rights_get(&entries[i]);
+
+		if (write) {
+			/* Need to check write permission. */
```

Comment:
Why are we spending instructions to validate the MPU fields?  We constructed and wrote those fields ourselves. The only way this trips the -EPERM is because of a bug.  At the very least it should be an __ASSERT, but really I think you could skip this.

Reply:
Hm?... because `arch_buffer_validate()` needs to validate the permission?

Reply:
Validate vs. what?  Again, nothing but code in this PR is generating those permission bits, these are all internal to the MPU layer, right?  Really this should be an assert, what's the circumstance at runtime where this could ever fail?